File: day-87/_312_Quick_Sort.py
# _312_Quick_Sort.py

# https://www.geeksforgeeks.org/problems/quick-sort/1?itm_source=geeksforgeeks&itm_medium=article&itm_campaign=practice_card

class Solution:
    def quickSort(self, arr, start, end):
        if start >= end:
            return

        if start < end:
            pI = self.partition(arr, start, end)
            self.quickSort(arr, start, pI - 1)
            self.quickSort(arr, pI + 1, end)

    def partition(self, arr, start, end):
        idx = start - 1
        pI = arr[end]

        for j in range(start, end):
            if arr[j] <= pI:
                idx += 1
                arr[j], arr[idx] = arr[idx], arr[j]

        idx += 1
        arr[end], arr[idx] = arr[idx], arr[end]
        return idx

# Input: arr[] = [4, 1, 3, 9, 7]
# Output: [1, 3, 4, 7, 9]

File: day-87/_312_Quick_Sort.cpp
// _312_Quick_Sort.cpp

// https://www.geeksforgeeks.org/problems/quick-sort/1?itm_source=geeksforgeeks&itm_medium=article&itm_campaign=practice_card

class Solution
{
public:
    void quickSort(vector<int> &arr, int start, int end)
    {
        if (start >= end)
        {
            return;
        }

        int pI = partition(arr, start, end);
        quickSort(arr, start, pI - 1);
        quickSort(arr, pI + 1, end);
    }

private:
    int partition(vector<int> &arr, int start, int end)
    {
        int idx = start - 1;
        int pI = arr[end];

        for (int j = start; j < end; j++)
        {
            if (arr[j] <= pI)
            {
                idx++;
                swap(arr[j], arr[idx]);
            }
        }

        idx++;
        swap(arr[end], arr[idx]);
        return idx;
    }
};

// Input: arr[] = [4, 1, 3, 9, 7]
// Output: [1, 3, 4, 7, 9]